[docs] [doc audit] Standalone expo module (#33159)
expo/expo: docs/pages/modules/use-standalone-expo-module-in-your-project.mdx

description: Learn how to use a standalone module created with create-expo-module in your project by using a monorepo or publishing the package to npm.
---

import { Grid01Icon } from '@expo/styleguide-icons/outline/Grid01Icon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> **The recommended way to create an Expo module** in an existing application is described in the [Expo Modules API: Get Started](/modules/get-started/) guide. This page explains two additional methods for using a module created with `create-expo-module` in an existing project:

Reviewer: ```suggestion
**The recommended way to create an Expo module** in an existing project is described in the [Expo Modules API: Get Started](/modules/get-started/) guide. This tutorial explains two additional methods for using a module created with `create-expo-module` in an existing project:
```

We always prefer to use the term "app" over "application" throughout the docs to reference an arbitrary app. However, in this case, the word project is more suitable since it references the app stage (development). We are also using the same word in the last line of this paragraph.

This page is also a tutorial so instead of using a generic term, let's be specific.

## Use a monorepo

Your project should use the following structure:

> - **apps/**: Store multiple projects here, including React Native apps.
> - **packages/**: Keep different packages used by your apps in this folder.
> - **package.json**: This is the root package file that contains the Yarn workspaces configuration.

Reviewer: ```suggestion
- **apps**: A directory to store multiple projects, including React Native apps.
- **packages**: A directory to keep different packages used by your apps.
- **package.json**: This is the root package file that contains the Yarn workspaces configuration.
```

Instead of using a slash at the end of directory names, let's remove it and spell it out. It is always helpful for users to read rather than see a symbol and wonder what it is meant for.

> **info** To learn how to configure your project as a monorepo, check out the [Working with monorepos](/guides/monorepos/) guide.

<Step label="1">

### Initialize a new module

Once you have set up the basic monorepo structure, create a new module using `create-expo-module` with the flag `--no-example` to skip creating the example app:

<Terminal cmd={['$ npx create-expo-module packages/expo-settings --no-example']} />

</Step>

<Step label="2">

### Set up workspace

> Now, configure `autolinking` so your apps can use the new module. Add the following block to the **package.json** file in each app under the **apps** directory:

Reviewer: ```suggestion
Configure `autolinking` so your apps can use the new module. Add the following block to the **package.json** file in each app inside the **apps** directory:
```

}
```

</Step>

<Step label="3">

### Run the module

Run one of your apps to ensure everything works. Then, start the TypeScript compiler in **packages/expo-settings** to watch for changes and rebuild the module's JavaScript:

<Terminal
  cmdCopy="cd packages/expo-settings && npm run build"
  cmd={['$ cd packages/expo-settings', '$ npm run build']}
/>

> Open another terminal, select an app from the **apps** directory, and run the `prebuild` command. Repeat these steps for each app in your monorepo to use the new module.

Reviewer: ```suggestion
Open another terminal window, select an app from the **apps** directory, and run the `prebuild` command with the `--clean` option. Repeat these steps for each app in your monorepo to use the new module.
```

  ]}
/>

> You can now use the module in your app. To test it, edit the **App.js** file in your app and render the text message from the `expo-settings` module:

Reviewer: The **App.js** file doesn't exist anymore in the default template.

```jsx App.js
import React from 'react';
import { Text, View } from 'react-native';
import * as Settings from 'expo-settings';

export default function App() {
  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Text>{Settings.hello()}</Text>
    </View>
  );
}
```

After this configuration, the app displays the text "Hello world! 👋".

</Step>

## Publish the module to npm

> Publish the module on npm, then install it as a dependency in your project.

Reviewer: ```suggestion
You can publish the module on npm and install it as a dependency in your project by following the steps below.
```

Not an instruction but an explanation of what people can do by following the steps below.

<Step label="1">

### Initialize a new module

Start by creating a new module with `create-expo-module`. Follow the prompts carefully, as you will publish this library, and choose a unique name for your npm package.

<Terminal cmd={['$ npx create-expo-module expo-settings']} />

</Step>

<Step label="2">

### Run the example project

> Run the example project to make sure everything is working. Then, start the TypeScript compiler to watch for changes and rebuild the JavaScript module.

Reviewer: ```suggestion
Run one of your apps to ensure everything works. Then, start the TypeScript compiler in the root of your project to watch for changes and rebuild the module's JavaScript:
```

Let's use the same phrase from the previous section here to use parallelism (which is helpful for end users when they are trying to skim through the information and implement it at the same time).
Also, this section previously didn't mention where to run this command. We should add that here.

  ]}
/>

</Step>

<Step label="3">

### Publish the package to npm

To publish your package to npm, you need an npm account. If you don't have one, create an account on [the npm website](https://www.npmjs.com/signup). After creating an account, log in by running the following command:

<Terminal cmd={['$ npm login']} />

Navigate to the root of your module, then run the following command to publish it:

<Terminal cmd={['$ npm publish']} />

Your module will now be published to npm and can be installed in other projects using `npm install`.

Apart from publishing your module to npm, you can use it in your project in the following ways:

- **Create a tarball**: Use `npm pack` to create a tarball of your module, then install it in your project by running `npm install /path/to/tarball`. This method is helpful for testing your module locally before publishing it or sharing it with others who don't have access to the npm registry.

> - **Run a local npm registry**: Use a tool like [Verdaccio](https://verdaccio.org/) to host a local npm registry. You can install your module from this registry, which is useful for managing internal packages within a company or organization.

Reviewer: ```suggestion
- **Run a local npm registry**: Use a tool such as [Verdaccio](https://verdaccio.org/) to host a local npm registry. You can install your module from this registry, which is useful for managing internal packages within a company or organization.
```

> - **Publish a private package**: [Use a private registry with EAS Builds](/build-reference/private-npm-packages/) to manage private modules securely.

Reviewer: ```suggestion
- **Publish a private package**: [Use a private registry with EAS Build](/build-reference/private-npm-packages/) to manage private modules securely.
```

Typo

</Step>

<Step label="4">

### Test the published module

To test the published module in a new project, create a new app and install the module as a dependency by running the following command:

<Terminal
  cmdCopy="npx create-expo-app my-app && cd my-app && npx expo install expo-settings"
  cmd={['$ npx create-expo-app my-app', '$ cd my-app', '$ npx expo install expo-settings']}
/>

> You can now use the module in your app! To test it, edit **App.js** and render the text message from **expo-settings**.

Reviewer: Again, creating a new app here, the default will be used and the entry point file will be **app/(tabs)/index.tsx**. Can you please update that here and in the previous section?
